src/Microsoft.AspNetCore.OData/EnableQueryAttribute.cs

```diff
@@ -38,6 +41,177 @@ public partial class EnableQueryAttribute : ActionFilterAttribute
+                else
+                {
+                    // In case where CLRType is missing, $count, $expand verifications cannot be done.
+                    // More importantly $expand required ODataQueryContext with clrType which cannot be done
+                    // If the model is untyped. Hence for such cases, letting the validation run post action.
+                    _queryValidationRunBeforeActionExecution = false;
+                    return;
+                }
+
+                _queryValidationRunBeforeActionExecution = true;
+            }
+            else
+            {
+                // For non-OData Json based controllers.
+                // For these cases few options are supported like IEnumerable<T>, Task<IEnumerable<T>>, T, Task<T>
+                // Other cases where we cannot determine the return type upfront, are not supported
+                // Like IActionResult, SingleResult. For such cases, the validation is run in OnActionExecuted
+                // When we have the result.
+                ControllerActionDescriptor controllerActionDescriptor = context.ActionDescriptor as ControllerActionDescriptor;
+
+                Type returnType = controllerActionDescriptor.MethodInfo.ReturnType;
+                Type elementType;
+
+                // For Task<> get the base object.
+                if (returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(Task<>))
+                {
+                    returnType = returnType.GetGenericArguments().First();
+                }
+
+                // For NetCore2.2+ new type ActionResult<> was created which encapculates IActionResult and T result.
+                // However we don't exactly have a version specific to NetCore2.2 (also at the time of writing this code
+                // 2.2 and 3.0 are both out of support), hence the code is made to work on NetCore3.1+ only.
+#if NETCOREAPP3_1 || NET5_0
+                if (returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(ActionResult<>))
+                {
+                    returnType = returnType.GetGenericArguments().First();
+                }
+#endif
+                if (TypeHelper.IsCollection(returnType))
+                {
+                    elementType = TypeHelper.GetImplementedIEnumerableType(returnType);
+                }
+                else if(TypeHelper.IsGenericType(returnType) && returnType.GetGenericTypeDefinition() == typeof(Task<>))
+                {
+                    elementType = returnType.GetGenericArguments().First();
+                }
+                else
+                {
+                    _queryValidationRunBeforeActionExecution = false;
+                    return;
+                }
+
+                IEdmModel edmModel = this.GetModel(
+                    elementType,
+                    request,
+                    controllerActionDescriptor);
+
+                if (controllerActionDescriptor != null)
```

Comment:
@RamjotSingh Under what scenarios would `controllerActionDescriptor` be null? The rest of the code in this method (including `GetModel` method) doesn't treat the variable like it can ever be null.

Reply:
Their are 2 derived types on top of ActionDescriptor https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.abstractions.actiondescriptor?view=aspnetcore-5.0 while the other one (RazorPages) is definitely not valid for OData actions, nothing stops a caller from adding the attribute on top of one. Hence the check. Additionally, more variants might be added in future.